File: templates/problems/18_bit_manipulation/03_counting_bits/solution.go
package counting_bits

// solveCountBits uses DP with the relation dp[i] = dp[i>>1] + (i&1). O(n) time, O(n) space.
func solveCountBits(n int) []int {
	dp := make([]int, n+1)
	for i := 1; i <= n; i++ {
		dp[i] = dp[i>>1] + (i & 1)
	}
	return dp
}

File: templates/problems/18_bit_manipulation/03_counting_bits/typescript/solution.ts
/**
 * Use DP with the relation dp[i] = dp[i>>1] + (i&1). O(n) time, O(n) space.
 */
export function solveCountBits(n: number): number[] {
    const dp = new Array(n + 1).fill(0);
    for (let i = 1; i <= n; i++) {
        dp[i] = dp[i >> 1] + (i & 1);
    }
    return dp;
}
